### ISSUE TITLE
BPPO seems to involve online evaluation when doing offline training

### ISSUE THREAD

**Issue by typoverflow** [Dragon-Zhuang/BPPO#1]
Hi, 
thanks for sharing the code. After digging into the implementation I found a potential issue that BPPO involves online evaluation when updating the `old_policy`. I'm afraid this adds more assumptions than we typically make in Offline RL scenarios where online evaluation is not accessible during training, which also makes the comparison unfair. What's more, what numbers are you reporting in the paper, the best score during training or the score of the policy from the last timestep? Is the online evaluation mentioned in the paper? 
https://github.com/Dragon-Zhuang/BPPO/blob/a84f310d02fa4a8a5bf31a3a6b67b6a1f8f7f201/main.py#L181-L190

Please correct me if there is any mis-understanding =)

**Comment by Dragon-Zhuang** [comment]
BPPO needs to compare the performance of two policies, but the classical off-policy evaluation methods are not ideal, that is, the training is extremely unstable. So, we adopt online evaluation unavoidably. 
At the current stage, most offline RL algorithms have to select hyperparameters or report best score during training (also including BPPO) through online evaluation. online evaluation is hard to avoid because reliable off-policy evaluation method is still missing. BPPO does rely on the online evaluation, so are other offline RL methods.

**Comment by typoverflow** [comment]
Sorry, but I don't think this justifies the use of online evaluation in BPPO. Online evaluation is analogous to testing model on the test dataset in supervised learning. In supervised learning, we typically select the model with the lowest validation loss on **the validation set** and report its performance on the test data. While in RL, the validation step corresponds to  offline policy evaluation, but not online evaluation as BPPO did in the white paper. The inclusion of online evaluation actually **leaks the otherwise forbidden information** into the policy optimization process, and BPPO exploits such benefits as it relies on online evaluation to update the old policy. 

For the second point you mentioned, actually this should break down into two things: the evaluation protocol and hyper-parameter selection strategy. 
> most offline RL algorithms have to select hyperparameters or report best score during training...online evaluation is hard to avoid because a reliable off-policy evaluation method is still missing. BPPO does rely on online evaluation, so are other offline RL methods.
1. **I don't think that reporting the best score during the entire training process is a convention**. I have checked the white paper as well as implementations of recent algorithms in this area, such as EDAC, IQL, XQL, CQL, COMBO, and MOPO, none of the papers and codes indicate that they reported the best score in the white paper. Among them, MOPO and COMBO explicitly mentioned that they report the score of the last checkpoint of the policy in the paper (Table 1 and Table 2 respectively). Do you have clear evidence that 'other offline rl methods' also report the best score? 
2. About how to appropriately select hyper-parameters, I agree that there is some missing consensus in this area on how much effort on tuning these parameters is acceptable. Recent offline algorithms such as [1] begin to incorporate OPE methods to decide on hyper-parameters in a fully offline manner. However, as I don't think this is a major concern, I'm not going into details on this topic. 

[1] RAMBO-RL: Robust Adversarial Model-Based Offline Reinforcement Learning.

**Comment by Dragon-Zhuang** [comment]
Most classical offline RL algorithms (including CQL, IQL, TD3+BC and so on) select hyperparameters through online evaluation. If these algorithms can be  tuned without online evaluation from scratch, please let offline RL community know. It will be an amazing finding. If not, they all **leaks the otherwise forbidden information**. Besides, some core hyperparameters of BPPO are directly from online PPO since no extra constrains or regularizations are introduced. This indicates when selecting hyperparameters, BPPO relies on less online evaluation than classical offline RL algorithms.

The biggest problem in offline RL is **No** off-policy evaluation method can replace online evaluation within the acceptable range. During implementation of BPPO, the problem confuses us a lot and we have no alternative but to chose online evaluation. In the future work, we will try more approaches to avoid the explicit use of online evaluation. And we will also add the results of the last checkpoint in future version.

Really appreciate for your discussions and comments =).

**Comment by vkurenkov** [comment]
I believe the main frustration is that the offline-RL papers usually report and compare the results of the last checkpoint. Reporting the best performance across the whole offline training progress usually leads to way higher scores (e.g. see our comparisons in [CORL](https://github.com/tinkoff-ai/CORL)).

For example, here are the results for locomotion datasets in your paper (which were taken from the original works as stated in the paper):
- BPPO - 83.4 (best score protocol)
- CQL - 77.6 (last score protocol)
- IQL - 76.9 (last score protocol)
- TD3+BC - 75.3 (last score protocol)

And here are the results for the same datasets reproduced in the [CORL](https://github.com/tinkoff-ai/CORL#best-scores):
- CQL - 85.4  (best score protocol)
- IQL - 86.4  (best score protocol)
- TD3+BC - 80.7 (best score protocol)

Moreover, I believe you can get even higher scores for all of the methods by increasing the number of evaluations. E.g., consider an extreme case where the model is evaluated after each gradient step.

Anyways, thanks for open-sourcing your code, this is very valuable.

**Comment by philippe-eecs** [comment]
@typoverflow @vkurenkov I think XQL also takes the max? I'm mostly referencing their antmaze eval curves which quickly peak and then drop. The last results don't match their table.

https://github.com/Div99/XQL/blob/main/offline/train_offline.py#L187

**Comment by vkurenkov** [comment]
@philippe-eecs this is worrisome as well, not sure whether they mentioned their evaluation protocol in the paper. Maybe you could open an issue in the XQL repo?

**Comment by ryanxhr** [comment]
> @typoverflow @vkurenkov I think XQL also takes the max? I'm mostly referencing their antmaze eval curves which quickly peak and then drop. The last results don't match their table.
> 
> https://github.com/Div99/XQL/blob/main/offline/train_offline.py#L187

Maybe you should try our SQL (https://github.com/ryanxhr/IVR), which is more robust than XQL and achieves similar performance, at least we didn't take the max:)

**Comment by philippe-eecs** [comment]
I've seen your paper! Very cool stuff, thanks for sharing.

I've lost interest in Offline RL as of late though... learning more about LLMs and Diffusion models!
